rclcpp/include/rclcpp/node.hpp

```diff
@@ -379,6 +379,11 @@ class Node : public std::enable_shared_from_this<Node>
 
   std::atomic_bool has_executor;
 
+  RCLCPP_PUBLIC
+  void
+  add_service(const rclcpp::service::ServiceBase::SharedPtr & service,
```

Comment:
If you actually read that post, it points out that copying shared pointers is more expensive compared to copying normal pointers or references, and that using a const reference only becomes a problem if you store the references in a vector or pass it to a function in another thread. Neither of which are happening here. I'd say that const reference is fine here, but you might still choose to not use const reference for other reasons (like the fact that it is a public interface which someone could misuse).

> In controlled circumstances you can pass the shared pointer by constant reference. Be sure that nobody is concurrently deleting the object, though this shouldn't be too hard if you're careful about to whom you give references.

I'd call this a controlled circumstance.

In this past we've solved issues before by removing the pass by reference of shared pointers, but in my opinion that's just using shared pointers to make the functions thread safe and generally doesn't have anything to do with shared pointers. We'd have the same problem if we were using raw pointers, i.e. someone calls this function in one thread and destroys the object (`Node`) in another thread. This is a general thread-safety issue, not a problem with how we're using shared pointers with const reference passing.